### PR TITLE
Changes for Actor Drop

### PR DESCRIPTION
**Pull request by JFranklin3000**
Closes [issue URL]

##### In this PR
- Modified drop for Megaform Combiners to drop Zords and Transformers
- Added drop to Vehicles and Zords that supports selecting passenger location
- Added drop for actors with the Zord Perk to drop Zords on to Characters

##### Testing
- Validate all of the drops in the PR work as expected. 
- Validate all other drops error.

### CODE REVIEW

File: templates/actor/parts/misc/system-actors.hbs [hunk 1]
```diff
@@ -0,0 +1,51 @@
+<div class="zord-combiners">
+  <div class="flexrow" style="gap: 6px;">
+    {{#each @root.actors as |attachedActor|}}
```
Comment (phildominguez) [inline templates/actor/parts/misc/system-actors.hbs:3]: Could we do something where if there are no actors present we show a message, like "Drop passengers here" or "Drop Zord Combiners here"?

File: lang/en.json [hunk 2]
```diff
@@ -65,6 +65,7 @@
     "ActorArmor": "Armor",
     "ActorConditioning": "Conditioning",
     "ActorDescription": "Description",
+    "ActorDropError": "The dropping of this actor type on to this target actor is not supported.",
```
Comment (phildominguez) [inline lang/en.json:68]:
```suggestion
    "ActorDropError": "The dropping of this Actor type onto this target Actor is not supported.",
```

File: module/helpers/config.mjs [hunk 3]
```diff
@@ -679,6 +679,15 @@ preLocalize("energonTypes");
...
 
+//Vehile Roles
+E20.vehicleRole = {
+  driver: "E20.VehicleRoleDriver",
+  gunner: "E20.VehicleRoleGunner",
+  passenger: "E20.VehicleRolePassenger",
+};
+preLocalize("vehicleRole");
+
+//Vehicle Traits
```
Comment (phildominguez) [inline module/helpers/config.mjs:690]:
```suggestion
// Vehicle Roles
E20.vehicleRoles = {
  driver: "E20.VehicleRoleDriver",
  gunner: "E20.VehicleRoleGunner",
  passenger: "E20.VehicleRolePassenger",
};
preLocalize("vehicleRole");

// Vehicle Traits
```

File: module/helpers/dialog.mjs [hunk 4]
```diff
@@ -18,6 +18,7 @@ export function rememberOptions(html) {
  * @returns {Object}  The dialog inputs and their submitted values
  */
 export function rememberSelect(html) {
+
```
Comment (phildominguez) [inline module/helpers/dialog.mjs:21]:
```suggestion
```

File: module/helpers/templates.mjs [hunk 5]
```diff
@@ -45,7 +46,7 @@ export const preloadHandlebarsTemplates = async function () {
     "systems/essence20/templates/actor/parts/misc/vehicle-crew.hbs",
     "systems/essence20/templates/actor/parts/misc/wealth-die.hbs",
     "systems/essence20/templates/actor/parts/misc/zord-common.hbs",
-    "systems/essence20/templates/actor/parts/misc/zordcombiner.hbs",
+
```
Comment (phildominguez) [inline module/helpers/templates.mjs:49]:
```suggestion
```

File: module/sheet-handlers/drop-handler.mjs [hunk 6]
```diff
@@ -118,25 +119,136 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
-  if (actor.type == 'megaformZord' && sourceActor.type == 'zord') {
-    const zordIds = duplicate(actor.system.zordIds);
-
-    // Can't contain duplicate Zords
-    if (!zordIds.includes(sourceActor.id)) {
-      zordIds.push(sourceActor.id);
-      await actor.update({
-        "system.zordIds": zordIds,
-      }).then(actorSheet.render(false));
+  let droppedActor = await fromUuid(data.uuid);
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:126]:
```suggestion
  const droppedActor = await fromUuid(data.uuid);
```

File: module/sheet-handlers/drop-handler.mjs [hunk 7]
```diff
@@ -118,25 +119,136 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
+
+    break;
+  case 'transformer':
+    if (droppedActor.type =='zord' && targetActor.system.canHaveZord || droppedActor.type == 'contact') {
+      setEntryAndAddActor (droppedActor, targetActor);
+    } else {
+      ui.notifications.error(game.i18n.localize('E20.ActorDropError'));
+    }
+
+    break;
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:169]: If the logic in these are all the same, you should be able to do something like:
```
case 'pony':
case 'powerRanger':
cast 'transformer':
  // logic here
```

File: module/sheet-handlers/drop-handler.mjs [hunk 8]
```diff
@@ -118,25 +119,136 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
+  case 'zord':
+    if (["giJoe", "npc", "pony", "powerRanger", "transformer"].includes(droppedActor.type)) {
+      _selectVehicleLocation(droppedActor, targetActor);
+    } else {
+      ui.notifications.error(game.i18n.localize('E20.ActorDropError'));
+    }
+
+    break;
+  default:
+    ui.notifications.error(game.i18n.localize('E20.ActorDropError'));
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:187]: I would make a variable before the switch (something like `dropIsValid`) that is initialized as false. Then, inside the `if` for each case set it to true. After the entire switch you can then do an if !dropIsValid-> notify, that way it reduces duplication.

File: module/sheet-handlers/drop-handler.mjs [hunk 9]
```diff
@@ -118,25 +119,136 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
-      await actor.update({
-        "system.zordIds": zordIds,
-      }).then(actorSheet.render(false));
+  let droppedActor = await fromUuid(data.uuid);
+  if (!droppedActor) return false;
+
+  switch (targetActor.type) {
+  case 'giJoe':
+    if (droppedActor.type =='zord' && targetActor.system.canHaveZord || droppedActor.type == 'contact') {
+      setEntryAndAddActor (droppedActor, targetActor);
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:132]:
```suggestion
      setEntryAndAddActor(droppedActor, targetActor);
```
Similar on the declarations below too

File: templates/dialog/vehicleRole-select.hbs
Comment (phildominguez) [inline templates/dialog/vehicleRole-select.hbs:1]: I would stick with kebab case, so vehicle-role-seelct.hbs

File: module/sheet-handlers/drop-handler.mjs [hunk 10]
```diff
@@ -118,25 +119,136 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
+      type: droppedActor.type,
+    };
+
+    if (["vehicle", "zord"].includes(targetActor.type)) {
+      entry['vehicleRole'] = options.vehicleRole;
+    }
+
+    return addActorIfUnique (droppedActor, targetActor, entry);
+
+  }
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:232]:
```suggestion
    return addActorIfUnique(droppedActor, targetActor, entry);
  }
```

File: module/sheet-handlers/drop-handler.mjs [hunk 11]
```diff
@@ -118,25 +119,136 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
+    const pathPrefix = "system.actors";
+    const key = createId(actors);
+
+    await targetActor.update({
+      [`${pathPrefix}.${key}`]: entry,
+    });
+
+    return key;
+  }
+
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:253]:
```suggestion
```

File: module/sheet-handlers/drop-handler.mjs [hunk 12]
```diff
@@ -118,25 +119,136 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
+    }
+
+    break;
+  default:
+    ui.notifications.error(game.i18n.localize('E20.ActorDropError'));
+
+    break;
   }
+
+  async function _selectVehicleLocation(droppedActor, targetActor) {
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:192]: Reminder for comment blocks 😃

File: module/sheet-handlers/power-ranger-handler.mjs [hunk 13]
```diff
@@ -1,17 +1,21 @@
...
  * Prepare Zords for MFZs.
  * @param {Actor} actor The Megaform Zord to prepare Zords for
  * @param {Object} context The actor data to prepare
- */
-export function prepareZords(actor, context) {
-  if (actor.type == 'megaformZord') {
-    let zords = [];
+*/
+export function prepareSystemActors(actor, context) {
+  if (Object.keys(actor.system.actors).length > 0) {
```
Comment (phildominguez) [inline module/sheet-handlers/power-ranger-handler.mjs:11]:
```suggestion
```

File: module/sheet-handlers/power-ranger-handler.mjs [hunk 14]
```diff
@@ -1,17 +1,21 @@
...
 
-    for (let zordId of actor.system.zordIds) {
-      zords.push(game.actors.get(zordId));
+    let actors = [];
```
Comment (phildominguez) [inline module/sheet-handlers/power-ranger-handler.mjs:12]:
```suggestion
    const actors = [];
```

File: module/sheet-handlers/power-ranger-handler.mjs [hunk 15]
```diff
@@ -1,17 +1,21 @@
+import { checkIsLocked } from "../helpers/actor.mjs";
+import { _getItemDeleteConfirmDialog } from "./listener-item-handler.mjs";
+
 /**
  * Prepare Zords for MFZs.
  * @param {Actor} actor The Megaform Zord to prepare Zords for
```
Comment (phildominguez) [inline module/sheet-handlers/power-ranger-handler.mjs:6]: Update these

File: module/sheet-handlers/power-ranger-handler.mjs [hunk 16]
```diff
@@ -20,11 +24,38 @@ export function prepareZords(actor, context) {
  * @param {Event} event The originating click event
  * @param {ActorSheet} actorSheet The ActorSheet whose Zord is being deleted
  */
```
Comment (phildominguez) [inline module/sheet-handlers/power-ranger-handler.mjs:26]: Update comment block

File: module/sheets/actor-sheet.mjs [hunk 17]
```diff
@@ -377,7 +377,7 @@ export class Essence20ActorSheet extends ActorSheet {
     html.find('.item-delete').click(ev => onItemDelete(ev, this));
 
     // Delete Zord from MFZ
```
Comment (phildominguez) [inline module/sheets/actor-sheet.mjs:379]: Update

File: module/sheets/actor-sheet.mjs [hunk 18]
```diff
@@ -81,7 +81,7 @@ export class Essence20ActorSheet extends ActorSheet {
     context.effects = prepareActiveEffectCategories(this.actor.effects);
 
     // Prepare Zords for MFZs
```
Comment (phildominguez) [inline module/sheets/actor-sheet.mjs:83]: Update

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: module/sheet-handlers/drop-handler.mjs (hunk 7 above)
Comment (JFranklin3000) [reply module/sheet-handlers/drop-handler.mjs:169]: changed

File: module/sheet-handlers/drop-handler.mjs (hunk 8 above)
Comment (JFranklin3000) [reply module/sheet-handlers/drop-handler.mjs:187]: changed.

File: module/sheet-handlers/drop-handler.mjs (hunk 12 above)
Comment (JFranklin3000) [reply module/sheet-handlers/drop-handler.mjs:192]: Added

File: module/sheets/actor-sheet.mjs (hunk 18 above)
Comment (JFranklin3000) [reply module/sheets/actor-sheet.mjs:83]: Updated

File: module/sheet-handlers/power-ranger-handler.mjs (hunk 16 above)
Comment (JFranklin3000) [reply module/sheet-handlers/power-ranger-handler.mjs:25]: Updated

File: module/sheet-handlers/power-ranger-handler.mjs (hunk 15 above)
Comment (JFranklin3000) [reply module/sheet-handlers/power-ranger-handler.mjs:6]: Updated

File: module/sheet-handlers/drop-handler.mjs [hunk 19]
```diff
@@ -113,30 +114,134 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
+    if (["giJoe", "npc", "pony", "powerRanger", "transformer"].includes(droppedActor.type)) {
+      _selectVehicleLocation(droppedActor, targetActor);
+      dropisValid = true;
+    }
+
+    break;
+  case 'zord':
+    if (["giJoe", "npc", "pony", "powerRanger", "transformer"].includes(droppedActor.type)) {
+      _selectVehicleLocation(droppedActor, targetActor);
+
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:158]:
```suggestion
```

File: module/sheet-handlers/drop-handler.mjs [hunk 20]
```diff
@@ -113,30 +114,134 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
+  if (!dropIsValid) {
+    ui.notifications.error(game.i18n.localize('E20.ActorDropError'));
+  }
+}
+
+/**
+ * Function to select where the actor is being seated
+ * @param {actor} droppedActor The actor that is being dropped
+ * @param {actor} targetActor The actor that is being dropped on to
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:172]:
```suggestion
 * @param {Actor} droppedActor The actor that is being dropped
 * @param {Actor} targetActor The actor that is being dropped on to
```

File: module/sheet-handlers/drop-handler.mjs [hunk 21]
```diff
@@ -113,30 +114,134 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
+      },
+    },
+  ).render(true);
+}
+
+/**
+ * Sets the entry value that will be stored in system.actors
+ * @param {actor} droppedActor Actor dropped on to another actor
+ * @param {actor} targetActor Actor that is being dropped on to
+ * @param {role} options An optional parameter for if a vehicle role has been selected
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:204]:
```suggestion
 * @param {Actor} droppedActor Actor dropped on to another actor
 * @param {Actor} targetActor Actor that is being dropped on to
 * @param {Objects} options An optional parameter for if a vehicle role has been selected
```

File: module/sheet-handlers/drop-handler.mjs [hunk 22]
```diff
@@ -113,30 +114,134 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
+ * @returns
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:205]: Missing what it returns

File: module/sheet-handlers/drop-handler.mjs [hunk 23]
```diff
@@ -113,30 +114,134 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
+  }
+
+  return addActorIfUnique(droppedActor, targetActor, entry);
+}
+
+/**
+ * function to add the actor in to the system.actors object
+ * @param {actor} droppedActor Actor dropped on to another actor
+ * @param {actor} targetActor Actor that is being dropped on to
+ * @param {object} entry The value to be written to the system.actors
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:226]:
```suggestion
 * Adds the dropped actor into system.actors
 * @param {Actor} droppedActor Actor dropped on to another actor
 * @param {Actor} targetActor Actor that is being dropped on to
 * @param {Object} entry The value to be written to the system.actors
```

File: module/sheet-handlers/drop-handler.mjs [hunk 24]
```diff
@@ -113,30 +114,134 @@ async function _onDropUpgrade(upgrade, actor, dropFunc) {
...
+ * @returns
```
Comment (phildominguez) [inline module/sheet-handlers/drop-handler.mjs:227]: Missing what it returns

File: templates/actor/parts/misc/system-actors.hbs [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+                {{selectOptions @root.config.vehicleRoles selected=current.vehicleRole}}
+              </select>
+            {{/ifEquals}}
+          {{/each}}
+        {{/inArray}}
+      </div>
+      {{/each}}
+    {{else}}
+      <div class="flexrow">
+        <input class="no-unlock" type="text" value="{{localize 'E20.ActorDrop'}}" readonly />
```
Comment (phildominguez) [inline templates/actor/parts/misc/system-actors.hbs:53]: I would just use (centered) plan text here

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: templates/actor/parts/misc/system-actors.hbs (hunk 1 above)
Comment (JFranklin3000) [reply templates/actor/parts/misc/system-actors.hbs]: added

File: module/sheet-handlers/drop-handler.mjs (hunk 22 above)
Comment (JFranklin3000) [reply module/sheet-handlers/drop-handler.mjs:204]: fixed

File: module/sheet-handlers/drop-handler.mjs (hunk 24 above)
Comment (JFranklin3000) [reply module/sheet-handlers/drop-handler.mjs:226]: fixed

File: templates/actor/parts/misc/system-actors.hbs (hunk 25 above)
Comment (JFranklin3000) [reply templates/actor/parts/misc/system-actors.hbs:53]: changed

**Review by phildominguez — APPROVED** [review]

**Pull request merged**